aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp

```diff
@@ -1128,5 +1130,136 @@ Tensor _sparse_csr_prod_cpu(const Tensor& input, IntArrayRef dims_to_reduce, boo
+
+  // Dim restrictions.
+  TORCH_CHECK(compressed.dim() == 2,
+      "spmm(", compressed_layout_str, ", Strided): sparse arguments which are not 2D are not supported.");
+  TORCH_CHECK(strided.dim() >= 2,
+      "spmm(", compressed_layout_str, ", Strided): expects strided inputs to be at least 2D.");
+
+  const auto m = compressed.sizes()[0];
+  const auto k = compressed.sizes()[1];
+  const auto n = strided.size(-1);
+  // Matrix product size compatibility.
+  TORCH_CHECK(strided.size(-2) == k,
+      "spmm(", compressed_layout_str, "Strided): argument sizes are not compatible for matrix multiplication. ",
+      "Got ", compressed_layout_str, ".sizes(-1) == ", k, " is not equal to ",
+      "Strided.sizes(-2) == ", strided.size(-2), ".");
+
+  TORCH_CHECK(result.layout() == kStrided,
+      "spmm_out(): out argument is expected to be strided.");
+  // We assume that result is properly resized.
+  auto result_expected_size = at::DimVector(strided.sizes().slice(0, strided.dim() - 2));
+  result_expected_size.push_back(m);
+  result_expected_size.push_back(n);
+  TORCH_CHECK(result.sizes() == result_expected_size,
+      "spmm_out(): out argument has wrong size. ",
+      "Expected (", result_expected_size, ") but got (", result.sizes(), ").");
+
+  auto values = compressed.values();
+
+  using Blocksize = std::array<int64_t, 2>;
+  // We refer to these as (b0, b1) in the comments below.
+  Blocksize blocksize = {1, 1};
+  if (compressed_layout == kSparseBsr) {
+    blocksize = {values.size(-2), values.size(-1)};
+  }
+
+  // NOTE: this function ALWAYS creates a view upon successful execution.
+  const auto tile_tensor = [compressed_layout](
+      const Tensor& t, Blocksize blocksize) -> Tensor {
+    if (compressed_layout == kSparseCsr || compressed_layout == kSparseCsc) {
+      return t.unsqueeze(-1).unsqueeze_(-1);
+    }
+    else {
+      const auto size_neg_2_blocked = t.size(-2) / blocksize[0];
+      const auto size_neg_1_blocked = t.size(-1) / blocksize[1];
+      auto tiled_sizes = at::DimVector(t.sizes().slice(0, t.dim() - 2));
+      tiled_sizes.push_back(size_neg_2_blocked);
+      tiled_sizes.push_back(blocksize[0]);
+      tiled_sizes.push_back(size_neg_1_blocked);
+      tiled_sizes.push_back(blocksize[1]);
+      return t.reshape(tiled_sizes).transpose(-3, -2);
+    }
+  };
+
+  // Note that sparse values are (..., b0, b1). This means that
+  // the strided input has to be "tilable" to (..., b1, x) with
+  // any x >= 1 such that all the shapes are (block) matrix product
+  // compatible. The matrix product will then have shape (..., b0, x).
+  // This in turn means the the result has to be "tilable" to
+  // (..., b0, x).
+  //
+  // These observations imply the following restrictions:
+  // 1. strided.size(-2) has to be divisible by b1.
+  // 2. result.size(-2) has to be divisible by b0.
+  // 3. both strided.size(-1) and result.size(-1)
+  //    have to be divisible by x.
+  //
+  // Restriction 2 is trivially satisfied.
```

Comment:
Restriction 1 is also trivially satisfied. 
For `A @ B -> C` to be well formed we will have verified that the sizes line up `(A(m, p) @ B(p, n) -> C(n, m)` 

If `A` is block sparse with `(b0, b1)` block-size then we already know that `p % b1== 0`.

Reply:
Agree, I overcomplicated :)